File: CR12/Lecture13/truncated_svd.m
function [U S V] = truncated_svd(A, accr)

	[U1 S1 V1] = svd(A, "ECO");
	
	[nrows ncolumns] = size(A);
	
	accr_2 = accr * accr;
   	trunc_error_2 = 0;
   	
   	r_trunc = min(nrows, ncolumns);

   	for k=r_trunc:-1:1

    		trunc_error_2 = trunc_error_2 + S1(k,k)*S1(k,k);

    		if trunc_error_2 > accr_2
    			r_trunc = k;
    			break
    		end

    	end
    	
    	U = U1(:,1:r_trunc);
    	S = S1(1:r_trunc,1:r_trunc);
    	V = V1(:,1:r_trunc);

end

File: CR12-2024/Lecture14/truncated_svd.py
import numpy as np
def truncated_svd(A, accr):
    nrows = A.shape[0]
    ncols = A.shape[1]

    U, S, VT = np.linalg.svd(A, full_matrices=False)

    accr2 = accr * accr
    trunc_error2 = 0.0

    r_trunc = 0

    for k in reversed(range(min(nrows, ncols))):
        trunc_error2 += S[k] * S[k]
        if trunc_error2 > accr2:
            r_trunc = k +1
            break

    U = U[:,0:r_trunc]
    S = S[0:r_trunc]
    VT = VT[0:r_trunc,:]
    return (U, S, VT)

